### PR TITLE
C++: Make sure there's a Instruction for each Expr

### PR DESCRIPTION
**Pull request by jbj**
> This change ensures that all `Expr`s (except parentheses) have a `TranslatedExpr` with a `getResult` that's one of its own instructions, not an instruction from one of its operands. This means that when we translate back and forth between `Expr` and `Instruction`, like in `DataFlow::exprNode`, we will not conflate `e` with `&e` or `... = e`.

This is my attempt to address the issue that causes confusing output in #1926. I think this change will be valuable for all future libraries that use the IR for implementation only and therefore must map back and forth between `Instruction` and `Expr`.

You might worry that this change will result in too many instructions, but that seems not to be the case. On MySQL, a C++ project, it increases the number of `Instruction`s from 7,431,574 to 7,890,523. Should we also add `ParenthesisExpr` to the set for the sake of consistency? That would get us up to 8,048,560 `Instructions`.

I first started a branch where I added the `CopyValue` instructions directly to each `TranslatedExpr` subclass. That was fine for simple cases, but it became complicated for the conditional ones; for example, `TranslatedCrementOperation` should only have a copy if it's C++ and is a prefix crement. It also led to unnecessary copies because there is no need to copy a result that has a `Load` on it. This made practically all the cases conditional, so i decided to centralize the copying with `TTranslatedResultCopy`. I've tried to write the `exprNeedsCopyIfNotLoaded` predicate in an object-oriented style, but [I couldn't convince QL that it was sound](https://semmle.slack.com/archives/C20TR4EAZ/p1569222685190800).

### CODE REVIEW

**Comment by dave-bartolomeo** [comment]
The code changes look fine, but I still have a few concerns about when we actually insert these `CopyValue` instructions. If I read the code right, we'll generate a `CopyValue` for the result of one of these expressions even if that expression is in a void context (e.g. `x = y;` generates a `CopyValue` that returns an lvalue to `x`). If the goal is make dataflow and other libraries interoperate with the IR better, do we really need "dead" instructions in the IR? I thought it was only expressions in the interior of the expression tree that really needed unique result instructions for the scenarios we care about.

Thanks for the instruction count numbers, BTW. I still don't think we need separate instructions for parentheses, but it's good to know that the cost isn't as bad as we expect if we do need them.

**Comment by jbj** [comment]
> If the goal is make dataflow and other libraries interoperate with the IR better, do we really need "dead" instructions in the IR?

That depends. When the goal is for the IR-based libraries to be drop-in replacements, then we want to mimic what the AST-based libraries do. We have qltests that look for all the flow between all the nodes in a program, and it'll currently show that there's flow from `source` to `x = source`.

When mapping back and forth between `Expr` and `Instruction`, the `getResult` values are pretty much all we have to go on. If we should not add `CopyValue` nodes for `=` in void context, then I suggest we make their `getResult` prediate have no result. That's better than giving it the same result as one of its subexpressions and risking that callers conflate them.

**Comment by dbartol** [comment]
> When mapping back and forth between `Expr` and `Instruction`, the `getResult` values are pretty much all we have to go on. If we should not add `CopyValue` nodes for `=` in void context, then I suggest we make their `getResult` prediate have no result. That's better than giving it the same result as one of its subexpressions and risking that callers conflate them.

You're referring to the `getResult()` predicate on `TranslatedExpr`, right? I think it would be fine for that predicate to not hold if the result is in a void context.

**Comment by jbj** [comment]
I'll give that a try.

**Comment by jbj** [comment]
I've checked that this change still has the intended effect on #1926. Still to do:
- Count instructions.
- Double-check performance.
- Fix new test failure.

**Comment by jbj** [comment]
@dbartol, this PR is ready for review again.

I've measured the instruction count increase this time on the aliased SSA form. I think it might have been on the raw form last time. The instruction count for MySQL on master is 9,051,892, and on this branch it's 9,364,789. So it's a smaller increase this time. Performance seems fine.

I've sampled some of the diff in test output relative to `master`, and the new copy instructions LGTM.

**Comment by jbj** [comment]
There were conflicts, so I've added a commit (76a3db9) that merges from `master` into this branch. The conflicts with #2008 were non-trivial, so please re-review my changes in the `TranslatedExpr` and `TranslatedCoreExpr` classes.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll [hunk 1]
```diff
@@ -2403,6 +2458,58 @@ class TranslatedErrorExpr extends TranslatedSingleInstructionExpr {
...
+}
+
+/**
+ * Holds if `expr` is immediately discarded. Such expressions do not need a
+ * `CopyValue` because it's unlikely that anyone is interested in their value.
+ */
+private predicate exprImmediatelyDiscarded(Expr expr) {
+  exists(ExprStmt s |
+    s = expr.getParent() and
+    not exists(StmtExpr se | s = se.getStmt().(Block).getLastStmt())
```
Comment (dbartol) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll]: Good catch.

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedExpr.qll]: I can't take credit for that; I copied it from https://github.com/Semmle/ql/blob/ee185ea92e60194b24df496aba8fd76e9143390f/cpp/ql/src/semmle/code/cpp/commons/VoidContext.qll#L26-L29

**Pull request merged**